### PR TITLE
contrib/net/http: add RTWithServiceName

### PR DESCRIPTION
**Pull request by CAFxX**
Allow to specify the service name in WrapClient and WrapRoundTripper, for symmetry with e.g. the [gRPC client middleware](https://godoc.org/gopkg.in/DataDog/dd-trace-go.v1/contrib/google.golang.org/grpc#WithServiceName).

### CODE REVIEW

**Comment by CAFxX** [comment]
note: the circleci failure seems to be unrelated:
```
# gopkg.in/DataDog/dd-trace-go.v1/contrib/go-redis/redis
contrib/go-redis/redis/redis.go:75:11: tc.Client.WrapProcess undefined (type *redis.Client has no field or method WrapProcess)
```

**Comment by cgilmour** [comment]
That CI failure was fixed recently.
Your branch point is slightly behind that, so it doesn't have the fix. Could you rebase it?

File: ddtrace/mocktracer/mocktracer.go [hunk 1]
```diff
@@ -40,13 +41,33 @@ type Tracer interface {
 // to activate the mock tracer. When your test runs, use the returned
 // interface to query the tracer's state.
 func Start() Tracer {
+	return StartWithOptions()
+}
+
+// StartWithOptions is the same as Start, but it allows to specify options.
+func StartWithOptions(opts ...StartOption) Tracer {
```
Comment (cgilmour) [inline ddtrace/mocktracer/mocktracer.go]: I think only one of `Start` and `StartWithOptions` is needed, eg: func `Start(opts ...StartOption) Tracer`

File: ddtrace/mocktracer/mocktracer.go [hunk 2]
```diff
@@ -40,13 +41,33 @@ type Tracer interface {
...
 	internal.SetGlobalTracer(&t)
 	internal.Testing = true
 	return &t
 }
 
+// StartOption is an option type to be passed to StartWithOptions
+type StartOption func(*mocktracer)
+
+// WithServiceName specifies the default service name
+func WithServiceName(serviceName string) StartOption {
```
Comment (cgilmour) [inline ddtrace/mocktracer/mocktracer.go]: The parameter here should be `name` for consistency with other `WithServiceName` functions.

File: contrib/net/http/roundtripper.go [hunk 3]
```diff
@@ -25,8 +25,11 @@ func (rt *roundTripper) RoundTrip(req *http.Request) (res *http.Response, err er
 		tracer.Tag(ext.HTTPMethod, req.Method),
 		tracer.Tag(ext.HTTPURL, req.URL.Path),
 	}
-	if rate := rt.cfg.analyticsRate; rate > 0 {
-		opts = append(opts, tracer.Tag(ext.EventSampleRate, rate))
+	if rt.cfg.analyticsRate > 0 {
```
Comment (gbbr) [inline contrib/net/http/roundtripper.go]: Why did you change this? Please revert the change.

File: contrib/net/http/roundtripper.go [hunk 4]
```diff
@@ -25,8 +25,11 @@ func (rt *roundTripper) RoundTrip(req *http.Request) (res *http.Response, err er
...
+		opts = append(opts, tracer.Tag(ext.EventSampleRate, rt.cfg.analyticsRate))
+	}
+	if rt.cfg.serviceName != "" {
```
Comment (gbbr) [inline contrib/net/http/roundtripper.go]: Let's use the same logic here, for consistency:
```go
if service := rt.cfg.serviceName; service != "" {
    // ...
```

File: ddtrace/mocktracer/mocktracer.go [hunk 5]
```diff
@@ -40,13 +41,33 @@ type Tracer interface {
 // to activate the mock tracer. When your test runs, use the returned
 // interface to query the tracer's state.
 func Start() Tracer {
+	return StartWithOptions()
```
Comment (gbbr) [inline ddtrace/mocktracer/mocktracer.go]: Please revert all the changes to the mocktracer. They don't add anything. The tests you wrote in `contrib/net/http` do not guarantee at all that the service name is or is not inherited, they simply test the mock tracer's behavior.

File: contrib/net/http/roundtripper_test.go [hunk 6]
```diff
@@ -129,3 +129,49 @@ func TestRoundTripperAnalyticsSettings(t *testing.T) {
...
+
+	t.Run("defaults", func(t *testing.T) {
+		mt := mocktracer.Start()
+		defer mt.Stop()
+
+		assertName(t, mt, nil)
+	})
+
+	t.Run("global", func(t *testing.T) {
+		mt := mocktracer.StartWithOptions(mocktracer.WithServiceName("service-name"))
```
Comment (gbbr) [inline contrib/net/http/roundtripper_test.go]: Instead, if you wish to test that the parent service _is_ or _is not_ inherited, simply call `RoundTrip` with an [`http.Request` context](https://godoc.org/net/http#Request.WithContext) that contains a span with a service name.

File: ddtrace/mocktracer/mocktracer.go (hunk 1 above)
Comment (CAFxX) [reply ddtrace/mocktracer/mocktracer.go]: That was my initial idea, then I decided to do a separate func because technically it's a breaking change. If you think it's ok I will add ...opts to Start.

File: contrib/net/http/roundtripper.go (hunk 3 above)
Comment (CAFxX) [reply contrib/net/http/roundtripper.go]: For symmetry with the new if, as this is easier to read. But sure it's ok, I can revert the change.

File: contrib/net/http/roundtripper.go (hunk 4 above)
Comment (CAFxX) [reply contrib/net/http/roundtripper.go]: Consistency and readability were the reasons I changed also the first one. If you really prefer that version I can change. I find your version less readable, but maybe it's just me.

File: contrib/net/http/roundtripper.go (hunk 3 above)
Comment (gbbr) [reply contrib/net/http/roundtripper.go]: Thanks!

File: contrib/net/http/roundtripper.go (hunk 4 above)
Comment (gbbr) [reply contrib/net/http/roundtripper.go]: Yes, please do. Probably `name` is a better name than `service`.

**Comment by gbbr** [comment]
@CAFxX we've actually recently discussed adding this option to `WrapRoundtripper`. There was a concern however that this might break other integrations setting the service in the "before func".

Furthermore, this also seems to make changes to the mock tracer, which is a separate issue.

Can we please close this PR and go through the regular contribution process? It's hard to follow the motivations behind it. All PRs should go through the process described in our [contribution guidelines](https://github.com/DataDog/dd-trace-go/blob/v1/CONTRIBUTING.md).

If you are still interested in this work, please open an issue and let's discuss it there before committing to any work. I'm confident we can find a solution that works.
